autolab/Autolab
# Callbacks for grace day usage

## Description
This PR adds the necessary callbacks that incrementally update watchlist instances for a course's grace day usage condition.

## Motivation and Context
This change is necessary for the metrics feature.

## How Has This Been Tested?
1. Add a grace day usage condition to the current metrics of the course of your preference. For initial testing, set grace day threshold to 1 and date to after testing date.
2. Create a new assignment for the course such that it is not yet past end date but already past due date.
3. Act as a student and make a submission to that assignment. Visit the endpoint to obtain current watchlist instances. Confirm that a new instance is created for that student.
4. Change the assignment's due date such that the student's use of grace days also changes. Revisit get_watchlist_instances and confirm that a new instance has replaced the old one with updated information.
5. Change the assignment's max grace days such that the student's use of grace days changes. Revisit the endpoint. Confirm a new instance has replaced the old one with updated information.
6. Change the course's late slack such that the student's use of grace days changes. Revisit get_watchlist_instances. Confirm a new instance has replaced the old one.
7. Change the course's grace days such that the student's use of grace days changes. Revisit get_watchlist_instances. Confirm a new instance has replaced the old one.
Note that in the above cases, if the change doesn't suffice to change a student's use of grace days, you should still observe that the previous watchlist instance is gone since it was archived, and if there is still a current one, its id should have been incremented.
8. Try with different combinations of the changes above. Compare the results of get_watchlist_instances with your expectation.

## Types of changes


## Checklist:


## Other issues / help required
See engineering channel for the callback condition issue. It has been fixed in this PR but whether it's a local or a more generic issue is yet to be confirmed.

If unsure, feel free to submit first and we'll help you along.
